service/matching/matching_engine_test.go

```diff
@@ -2499,6 +2497,100 @@ func (s *matchingEngineSuite) TestUnknownBuildId_Match() {
 	wg.Wait()
 }
 
+func (s *matchingEngineSuite) TestUnknownBuildId_Demoted_Match() {
+	namespaceId := namespace.ID(uuid.New())
+	tq := "makeToast"
+	unknown := "unknown"
+	build1 := "build1"
+
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	s.mockMatchingClient.EXPECT().UpdateWorkerBuildIdCompatibility(gomock.Any(), &matchingservice.UpdateWorkerBuildIdCompatibilityRequest{
+		NamespaceId: namespaceId.String(),
+		TaskQueue:   tq,
+		Operation: &matchingservice.UpdateWorkerBuildIdCompatibilityRequest_PersistUnknownBuildId{
+			PersistUnknownBuildId: unknown,
+		},
+	}).Return(&matchingservice.UpdateWorkerBuildIdCompatibilityResponse{}, nil).AnyTimes() // might get called again on dispatch from spooled
+
+	// add a task for an unknown build id, will get redirected to guessed set
+	_, err := s.matchingEngine.AddWorkflowTask(ctx, &matchingservice.AddWorkflowTaskRequest{
+		NamespaceId:      namespaceId.String(),
+		Execution:        &commonpb.WorkflowExecution{RunId: "run", WorkflowId: "wf"},
+		ScheduledEventId: 123,
+		TaskQueue:        &taskqueuepb.TaskQueue{Name: tq, Kind: enumspb.TASK_QUEUE_KIND_NORMAL},
+		VersionDirective: &taskqueue.TaskVersionDirective{
+			Value: &taskqueue.TaskVersionDirective_BuildId{
+				BuildId: unknown,
+			},
+		},
+	})
+	s.NoError(err)
+	// allow taskReader to finish starting dispatch loop so we don't get an extra load
+	time.Sleep(10 * time.Millisecond)
+
+	// unload base and versioned tqm. note: unload versioned first since versioned taskReader
+	// tries to load base for dispatching.
+	id := newTestTaskQueueID(namespaceId, tq, enumspb.TASK_QUEUE_TYPE_WORKFLOW)
+	verId := newTaskQueueIDWithVersionSet(id, hashBuildId(unknown))
+	verTqm, err := s.matchingEngine.getTaskQueueManager(ctx, verId, normalStickyInfo, false)
+	s.NoError(err)
+	s.NotNil(verTqm)
+	s.matchingEngine.unloadTaskQueue(verTqm)
+	// allow taskReader goroutines time to exit
+	time.Sleep(10 * time.Millisecond)
+
+	// unload base
+	baseTqm, err := s.matchingEngine.getTaskQueueManager(ctx, id, normalStickyInfo, false)
+	s.NoError(err)
+	s.NotNil(baseTqm)
+	s.matchingEngine.unloadTaskQueue(baseTqm)
+	// allow taskReader goroutines time to exit
+	time.Sleep(10 * time.Millisecond)
+
+	// both are now unloaded. change versioning data to merge unknown into another set.
+	clock := hlc.Zero(1)
+	userData := &persistencespb.TaskQueueUserData{
+		Clock: &clock,
+		VersioningData: &persistencespb.VersioningData{
+			VersionSets: []*persistencespb.CompatibleVersionSet{
+				{
+					// make "unknown" the demoted one to test demoted set loading.
+					// it works the other way too but doesn't test anything new.
```

Comment:
So, there's something else that tests loading demoted set ids, but it doesn't test loading unknown demoted set ids? Can you explain what is new specifically that this is testing? I'm not sure what an unknown build id even means in this case because it looks like it is known if it's in the version compatibility data here.

Reply:
It's unknown in the sense that it snuck into the versioning data via an AddTask request as opposed to the "front door" where a user explicitly adds it via the UpdateTaskQueueCompatibility API.

Reply:
I mean that if `hash(unknown)` was the primary set id (not demoted), then it would be the one that the poll would be redirected to, so it would just be testing normal versioned task dispatch. the point is that the task is living in the db in one version set id the poll is redirected to another version set id, so to get them to match we need to 1. load the demoted one and 2. redirect its tasks when they come out of taskReader. 2 is already done and 1 is this PR.